app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp

```diff
@@ -1,214 +1,702 @@
+        return AndroidStatus::Ok;
     }
 
-    void GraphicBufferProducer::QueueBuffer(Parcel &in, Parcel &out) {
-        struct Data {
-            u32 slot;
-            u64 timestamp;
-            u32 autoTimestamp;
-            ARect crop;
-            u32 scalingMode;
-            u32 transform;
-            u32 stickyTransform;
-            u64 _unk0_;
-            u32 swapInterval;
-            std::array<nvdrv::Fence, 4> fence;
-        } &data = in.Pop<Data>();
-
-        auto buffer{queue.at(data.slot)};
-        buffer->status = BufferStatus::Queued;
-
-        buffer->texture->SynchronizeHost();
-        state.gpu->presentation.Present(buffer->texture);
-        queue.at(data.slot)->status = BufferStatus::Free;
-        state.gpu->presentation.bufferEvent->Signal();
-
-        struct {
-            u32 width;
-            u32 height;
-            u32 _pad_[3];
-        } output{
-            .width = buffer->gbpBuffer.width,
-            .height = buffer->gbpBuffer.height,
-        };
-        out.Push(output);
-
-        state.logger->Debug("Timestamp: {}, Auto Timestamp: {}, Crop: [T: {}, B: {}, L: {}, R: {}], Scaling Mode: {}, Transform: {}, Sticky Transform: {}, Swap Interval: {}, Slot: {}", data.timestamp, data.autoTimestamp, data.crop.top, data.crop.bottom, data.crop.left, data.crop.right, data.scalingMode, data.transform, data.stickyTransform, data.swapInterval, data.slot);
+    AndroidStatus GraphicBufferProducer::DetachBuffer(i32 slot) {
+        std::scoped_lock lock(mutex);
+        if (slot < 0 || slot >= queue.size()) [[unlikely]] {
+            state.logger->Warn("#{} was out of range", slot);
+            return AndroidStatus::BadValue;
+        }
+
+        auto &bufferSlot{queue[slot]};
+        if (bufferSlot.state != BufferState::Dequeued) [[unlikely]] {
+            state.logger->Warn("#{} was '{}' instead of being dequeued", slot, ToString(bufferSlot.state));
+            return AndroidStatus::BadValue;
+        } else if (!bufferSlot.wasBufferRequested) [[unlikely]] {
+            state.logger->Warn("#{} was detached prior to being requested", slot);
+            return AndroidStatus::BadValue;
+        }
+
+        bufferSlot.state = BufferState::Free;
+        bufferSlot.frameNumber = std::numeric_limits<u32>::max();
+        bufferSlot.graphicBuffer = nullptr;
+
+        bufferEvent->Signal();
+
+        state.logger->Debug("#{}", slot);
+        return AndroidStatus::Ok;
     }
 
-    void GraphicBufferProducer::CancelBuffer(Parcel &in) {
-        u32 slot{in.Pop<u32>()};
-        //auto fences{in.Pop<std::array<nvdrv::Fence, 4>>()};
+    AndroidStatus GraphicBufferProducer::DetachNextBuffer(std::optional<GraphicBuffer> &graphicBuffer, std::optional<AndroidFence> &fence) {
+        std::scoped_lock lock(mutex);
+        auto bufferSlot{queue.end()};
+        for (auto it{queue.begin()}; it != queue.end(); it++) {
```

Comment:
Could find_if work here?

Reply:
See how it doesn't `break` after finding one, it needs to find the oldest `Free` slot rather than just any `Free` slot in the queue

Reply:
So reversed find_if?

Reply:
What if the first slot is the oldest? The queue isn't ordered by age